### DISCUSSION TITLE
Improve health check (Circuit Breakers): Allow fallback if upstream threshold reached

### DISCUSSION THREAD

**Discussion by Ehekatl** [Kong/kong#8909] [Ideas and feature requests]
Passive health check can be good in most of the case, however, it doesn't works well in the following scenario:
- Case 1: Flapping/Unavailable upstream, sometimes it's easy to trigger passive health check when some of the traffic goes wrong, but active health check will work and put the target back, without luck you may got a completely "dead" upstream, and cause production issue.
- Case 2: Aggressive Circuit Breakers, sometimes you break the traffic not because upstream can not handle it, but simply because it doesn't handle it fast enough due to latency issue, it helps to maximize the E2E latency when you configure very aggressive passive check, and for sure you don't want the traffic being completely dropped while they can still serve the traffic.
- Case 3: Mis-configuration or request spike, we don't want real traffic being blocked by random error.

I think introduce a few option on upstream threshold would be nice:
- `healthchecks.fallback` default false, when set to true, all failed target will consider healthy if pass the `healthchecks.threshold`
- `healthchecks.fallback_timeout` when healthcheck fallback is working, we might want a relatively short timeout in order to save kong it itself from being blocked by upstream，default value can be 1s, and let user decide what suits the best.
Pros: it solves all the issue mentioned above
Cons: passing all the traffic to backend while it's really overload might cause connection hanging, but I think it's acceptable, in that case, you will have to look into the issue no matter what, you can halt the faulty traffic if you really don't want. It's a better trade off for most of the cases.

I'm willing to contribute PR, but balancer logic it's at the very core part of Kong (and historically introduced lots of issue around it)
So hope this feature could be officially support.

**Comment by Tieske** [comment]
> * Case 1: Flapping/Unavailable upstream, sometimes it's easy to trigger passive health check when some of the traffic goes wrong, but active health check will work and put the target back, without luck you may got a completely "dead" upstream, and cause production issue.

I fail to see what is wrong in this case? a "Flapping/Unavailable upstream" is a production issue to begin with. Kong cannot prevent these issues from happening, or make them go away. It can only deal with them in a specific way. And that's exactly what passive/active healthchecks do.

> 
> * Case 2: Aggressive Circuit Breakers, sometimes you break the traffic not because upstream can not handle it, but simply because it doesn't handle it fast enough due to latency issue, it helps to maximize the E2E latency when you configure very aggressive passive check, and for sure you don't want the traffic being completely dropped while they can still serve the traffic.

Availability means both a service NOT responding as well as a service responding too late (high latency). If that is something that should be tolerated, then the checks should be configured accordingly. Relax the constraints on your health checks. If you cannot tolerate it, increase capacity on your upstreams.

> * Case 3: Mis-configuration or request spike, we don't want real traffic being blocked by random error.

No extra option can shield you from this. What if the mis-configuration-options get mis-configured?

**Reply by Ehekatl** [reply]
We use kong intensively on production workloads. These scenario comes from real production cases.
Let me clarify them:
- For case 1, there is a LOT cases, devs mis-configured passive health check, the entire service just goes down while those error could actually being ignored. Ya it's configuration issue, but let gateway be smarter is easier than let people never make mistake.
- For case 2, the latency from different target might be different, but instead of evenly distribute the traffic, it's better to distribute it around latency, like configure active check timeout at 100ms, so it filtered out any high latency targets, with option like `healthchecks.fallback` it can guarantee such trick won't break when all targets goes to 200ms avg while it's totally acceptable. 
- For case 3, this option is only meant for people who knows what they are doing, and it's common in Cloud, cloud ALB like aws, will route traffic to all targets by default if all health check failed.
